File: 953. Verifying an Alien Dictionary.py
class Solution:
    def isAlienSorted(self, words: List[str], order: str) -> bool:
        '''
        The problem asks us to find out the two words follows a <= b
        so we need to check if all char in the two words;
        one case is if the shorter one is the same as the longer's prefix, then the shorter is smaller
        '''
        ord_index = {c:i for i,c in enumerate(order)}
        
        for i in range(len(words)-1):
            word1, word2 = words[i], words[i+1]
            for i in range(min(len(word1), len(word2))):
                if word1[i] != word2[i]:
                    if ord_index[word1[i]] > ord_index[word2[i]]:
                        return False
                    break
            else:
                if len(word1) > len(word2):
                    return False
        return True

File: 953. Verifying an Alien Dictionary.java
class Solution {
    public boolean isAlienSorted(String[] words, String order) {
        HashMap<Character,Integer> map = new HashMap<>();
        for (int i=0; i<order.length(); i++) {
            map.put(order.charAt(i), i);
        }
        for (int i=0; i<words.length-1; i++) {
            int pa = 0, pb = 0;
            String a = words[i];
            String b = words[i+1];
            while (pa < a.length() && pb < b.length()) {
                int temp = map.get(a.charAt(pa)) - map.get(b.charAt(pb));
                if (temp > 0) {
                    return false;
                } else if (temp < 0) {
                    break;
                } else {
                    pa++; pb++;
                }
            }
            if (pb == b.length() && pa != a.length()) {
                return false;
            }
        }
        return true;
    }
}